File: samples/java/ZoneWorld/Server/src/main/java/systems/zlink/samples/zoneworld/server/gateway/RelocationProbeService.java
package systems.zlink.samples.zoneworld.server.gateway;

import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Component;

import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkSpotManager;
import systems.zlink.samples.zoneworld.shared.Messages;
import systems.zlink.samples.zoneworld.shared.ZoneWorldNames;

import java.util.List;
import java.util.concurrent.CompletionStage;

@Component
@ConditionalOnProperty(prefix = "sample", name = "role", havingValue = "gateway")
public final class RelocationProbeService {
    private static final List<List<String>> ADJACENT =
            List.of(
                    List.of("zone-nw", "zone-ne"), List.of("zone-nw", "zone-sw"),
                    List.of("zone-ne", "zone-se"), List.of("zone-sw", "zone-se"));
    private final ZLinkSpotManager spots;
    private final ZLinkActorManager actors;

    public RelocationProbeService(ZLinkSpotManager spots, ZLinkActorManager actors) {
        this.spots = spots;
        this.actors = actors;
    }

    public CompletionStage<Messages.RelocationPairRes> selectPair() {
        return selectPair(0);
    }

    private CompletionStage<Messages.RelocationPairRes> selectPair(int index) {
        if (index >= ADJACENT.size())
            return java.util.concurrent.CompletableFuture.completedFuture(
                    new Messages.RelocationPairRes("", "", "", "", "NoCrossNodeAdjacentPair"));
        List<String> pair = ADJACENT.get(index);
        return spots.find(pair.get(0))
                .thenCombine(
                        spots.find(pair.get(1)),
                        (source, target) -> {
                            if (source.isPresent()
                                    && target.isPresent()
                                    && !source.get().nodeRid().equals(target.get().nodeRid())) {
                                return new Messages.RelocationPairRes(
                                        pair.get(0),
                                        pair.get(1),
                                        source.get().nodeRid().toString(),
                                        target.get().nodeRid().toString(),
                                        null);
                            }
                            return null;
                        })
                .thenCompose(
                        found ->
                                found != null
                                        ? java.util.concurrent.CompletableFuture.completedFuture(
                                                found)
                                        : selectPair(index + 1));
    }

    public CompletionStage<Messages.ActorLocationProbeRes> findActor(String actorId) {
        return actors.find(actorId)
                .thenApply(
                        found ->
                                found.map(
                                                actor ->
                                                        new Messages.ActorLocationProbeRes(
                                                                actor.actorId(),
                                                                actor.objectGeneration(),
                                                                actor.nodeRid().toString(),
                                                                null))
                                        .orElseGet(
                                                () ->
                                                        new Messages.ActorLocationProbeRes(
                                                                actorId, 0, "", "ActorNotFound")));
    }

    public CompletionStage<Messages.FreshActorProbeRes> createFresh(String actorId) {
        return actors.getOrCreate(actorId, ZoneWorldNames.PLAYER_ACTOR_TYPE)
                .inMesh(ZoneWorldNames.MESH)
                .request(ZLinkMessage.empty())
                .submit()
                .thenApply(
                        result -> {
                            ActorRef actor =
                                    result instanceof ZLinkActorCreateResult.Created created
                                            ? created.actor()
                                            : result
                                                            instanceof
                                                            ZLinkActorCreateResult.Existing existing
                                                    ? existing.actor()
                                                    : null;
                            return actor == null
                                    ? new Messages.FreshActorProbeRes(
                                            actorId, 0, "", "ActorCreateRejected")
                                    : new Messages.FreshActorProbeRes(
                                            actor.actorId(),
                                            actor.objectGeneration(),
                                            actor.nodeRid().toString(),
                                            null);
                        });
    }
}

File: samples/kotlin/ZoneWorld/Server/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/server/gateway/RelocationProbeService.kt
package systems.zlink.samples.kotlin.zoneworld.server.gateway

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Component
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.spots.ZLinkSpotManager
import systems.zlink.samples.kotlin.zoneworld.shared.Messages
import systems.zlink.samples.kotlin.zoneworld.shared.ZoneWorldNames

@Component
@ConditionalOnProperty(prefix = "sample", name = ["role"], havingValue = "gateway")
class RelocationProbeService(
    private val spots: ZLinkSpotManager,
    private val actors: ZLinkActorManager,
) {
    private val adjacent =
        listOf(
            listOf("zone-nw", "zone-ne"),
            listOf("zone-nw", "zone-sw"),
            listOf("zone-ne", "zone-se"),
            listOf("zone-sw", "zone-se"),
        )

    fun selectPair(index: Int = 0): CompletionStage<Messages.RelocationPairRes> {
        if (index >= adjacent.size)
            return CompletableFuture.completedFuture(
                Messages.RelocationPairRes("", "", "", "", "NoCrossNodeAdjacentPair")
            )
        val pair = adjacent[index]
        return spots
            .find(pair[0])
            .thenCombine(spots.find(pair[1])) { source, target ->
                if (
                    source.isPresent &&
                        target.isPresent &&
                        source.get().nodeRid() != target.get().nodeRid()
                ) {
                    Messages.RelocationPairRes(
                        pair[0],
                        pair[1],
                        source.get().nodeRid().toString(),
                        target.get().nodeRid().toString(),
                    )
                } else null
            }
            .thenCompose { found ->
                if (found != null) CompletableFuture.completedFuture(found)
                else selectPair(index + 1)
            }
    }

    fun findActor(actorId: String): CompletionStage<Messages.ActorLocationProbeRes> =
        actors.find(actorId).thenApply { found ->
            if (found.isPresent)
                found.get().let {
                    Messages.ActorLocationProbeRes(
                        it.actorId(),
                        it.objectGeneration(),
                        it.nodeRid().toString(),
                    )
                }
            else Messages.ActorLocationProbeRes(actorId, 0, "", "ActorNotFound")
        }

    fun createFresh(actorId: String): CompletionStage<Messages.FreshActorProbeRes> =
        actors
            .getOrCreate(actorId, ZoneWorldNames.PLAYER_ACTOR_TYPE)
            .inMesh(ZoneWorldNames.MESH)
            .request(ZLinkMessage.empty())
            .submit()
            .thenApply { result ->
                val actor =
                    when (result) {
                        is ZLinkActorCreateResult.Created -> result.actor()
                        is ZLinkActorCreateResult.Existing -> result.actor()
                        is ZLinkActorCreateResult.Rejected -> null
                    }
                actor?.let {
                    Messages.FreshActorProbeRes(
                        it.actorId(),
                        it.objectGeneration(),
                        it.nodeRid().toString(),
                    )
                } ?: Messages.FreshActorProbeRes(actorId, 0, "", "ActorCreateRejected")
            }
}
